File: app/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/fragments/investigation/evidence/children/solo/views/SanitySeekBarView.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.evidence.children.solo.views;

import android.annotation.SuppressLint;
import android.content.Context;
import android.text.Html;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.Spanned;
import android.util.AttributeSet;
import android.util.Log;
import android.widget.SeekBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatSeekBar;
import androidx.appcompat.widget.AppCompatTextView;

import com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.evidence.data.SanityData;
import com.TritiumGaming.phasmophobiaevidencepicker.data.utilities.FontUtils;
import com.TritiumGaming.phasmophobiaevidencepicker.data.utilities.FormatterUtils;

public class SanitySeekBarView extends AppCompatSeekBar {

    private SanityData sanityData;

    public SanitySeekBarView(@NonNull Context context) {
        super(context);
    }

    public SanitySeekBarView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public SanitySeekBarView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void init(SanityData sanityData, AppCompatTextView sanityPercentTextView){
        this.sanityData = sanityData;

        setProgress((int)sanityData.getSanityActual());

        setOnSeekBarChangeListener(new OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {

                    sanityData.setProgressManually(progress);
                    sanityData.tick();

                    /*sanityPercentTextView.setText(sanityData.toPercentString());*/
                    sanityPercentTextView.setText(formatSanityPercent());
                    sanityPercentTextView.invalidate();

                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });

        sanityPercentTextView.setText(formatSanityPercent());
    }

    @SuppressLint("DefaultLocale")
    public String formatSanityPercent() {
        String nbsp = "\u00A0";
        String input = String.format("%3d", Integer.parseInt(sanityData.toPercentString()));

        StringBuilder output = new StringBuilder();
        int i = 0;
        for(; i < input.length(); i++) {
            if(input.charAt(i) != '0') { break; }
            output.append(nbsp);
        }
        for(; i < input.length(); i++) {
            output.append(input.charAt(i));
        }
        output.append("%");

        return output.toString();
    }

    public void updateProgress() {
        setProgress((int) sanityData.getSanityActual());
        formatSanityPercent();
        invalidate();
    }

    public void resetProgress() {
        setProgress(0);
        formatSanityPercent();
        invalidate();
    }

    /*
    @Override
    public void invalidate() {
        super.invalidate();

        sanityPercentTextView.setText(formatSanityPercent());
    }
    */
}

File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/views/investigation/sanity/SanitySeekBarView.kt
package com.TritiumGaming.phasmophobiaevidencepicker.views.investigation.sanity

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.widget.SeekBar
import androidx.appcompat.widget.AppCompatSeekBar
import androidx.appcompat.widget.AppCompatTextView
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.evidence.data.SanityModel
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.EvidenceViewModel

class SanitySeekBarView : AppCompatSeekBar {

    private lateinit var evidenceViewModel: EvidenceViewModel
    private var sanityPercentTextView: AppCompatTextView? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    fun init(evidenceViewModel: EvidenceViewModel, sanityPercentTextView: AppCompatTextView?) {
        this.evidenceViewModel = evidenceViewModel
        this.sanityPercentTextView = sanityPercentTextView

        progress = evidenceViewModel.sanityData?.sanityActual?.toInt() ?: 0

        setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    evidenceViewModel.sanityData?.setProgressManually(progress.toLong())
                    evidenceViewModel.sanityData?.tick()

                    /*sanityPercentTextView.setText(sanityData.toPercentString());*/
                    if (sanityPercentTextView != null) {
                        sanityPercentTextView.text = formatSanityPercent()
                        //sanityPercentTextView.invalidate();
                    }
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
            }
        })

        if (sanityPercentTextView != null) {
            sanityPercentTextView.text = formatSanityPercent()
        }
    }

    @SuppressLint("DefaultLocale")
    fun formatSanityPercent(): String {
        val nbsp = "\u00A0"

        var percentStr = evidenceViewModel.sanityData?.toPercentString() ?: "NA"
        percentStr = percentStr.replace(nbsp, "").trim { it <= ' ' }

        var percentNum = 100
        try {
            percentNum = percentStr.toInt()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }

        val input = String.format("%3d", percentNum)

        val output = StringBuilder()
        var i = 0
        while (i < input.length) {
            if (input[i] != '0') {
                break
            }
            output.append(nbsp)
            i++
        }
        while (i < input.length) {
            output.append(input[i])
            i++
        }
        output.append("%")

        return output.toString()
    }

    fun updateProgress() {
        progress = evidenceViewModel.sanityData?.sanityActual?.toInt() ?: 0
        if (sanityPercentTextView != null) {
            sanityPercentTextView!!.text = formatSanityPercent()
        }
        invalidate()
    }

    fun resetProgress() {
        progress = 0
        if (sanityPercentTextView != null) {
            sanityPercentTextView!!.text = formatSanityPercent()
        }
        invalidate()
    } /*
    @Override
    public void invalidate() {
        super.invalidate();

        sanityPercentTextView.setText(formatSanityPercent());
    }
    */
}
